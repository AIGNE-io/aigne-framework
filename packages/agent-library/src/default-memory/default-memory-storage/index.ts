import type { AgentInvokeOptions, Context, Memory } from "@aigne/core";
import type { PromiseOrValue } from "@aigne/core/utils/type-utils.js";
import { initDatabase } from "@aigne/sqlite";
import { desc, eq, type InferSelectModel, isNull, sql } from "drizzle-orm";
import type { SqliteRemoteDatabase } from "drizzle-orm/sqlite-proxy";
import { v7 } from "uuid";
import { stringify } from "yaml";
import { MemoryStorage } from "../storage.js";
import { migrate } from "./migrate.js";
import { Memories } from "./models/memory.js";

const DEFAULT_MAX_MEMORY_COUNT = 10;

export interface DefaultMemoryStorageOptions {
  url?: string;
  getSessionId?: (context: Context) => PromiseOrValue<string>;
  enableFTS?: boolean;
}

export class DefaultMemoryStorage extends MemoryStorage {
  constructor(public options?: DefaultMemoryStorageOptions) {
    super();
  }

  private _db: ReturnType<typeof this.initSqlite> | undefined;

  private async initSqlite() {
    const db = (await initDatabase({ url: this.options?.url })) as unknown as SqliteRemoteDatabase;

    await migrate(db);

    return db;
  }

  get db() {
    this._db ??= this.initSqlite();

    return this._db;
  }

  private convertMemory(m: InferSelectModel<typeof Memories>): Memory {
    return {
      id: m.id,
      sessionId: m.sessionId,
      content: m.content,
      createdAt: m.createdAt.toISOString(),
    };
  }

  async search(
    query: { search?: string; limit?: number },
    { context }: AgentInvokeOptions,
  ): Promise<{ result: Memory[] }> {
    const { limit = DEFAULT_MAX_MEMORY_COUNT } = query;

    const sessionId = (await this.options?.getSessionId?.(context)) ?? null;

    const db = await this.db;

    const memories =
      this.options?.enableFTS && query.search
        ? await db
            .select()
            .from(Memories)
            .innerJoin(sql`Memories_fts`, sql`Memories_fts.id = ${Memories.id}`)
            .where(sql`Memories_fts MATCH ${sql.param(this.segment(query.search).join(" OR "))}`)
            .orderBy(sql`bm25(Memories_fts)`)
            .limit(limit)
            .execute()
            .then((rows) => rows.map((row) => row.Memories))
        : await db
            .select()
            .from(Memories)
            .where(sessionId ? eq(Memories.sessionId, sessionId) : isNull(Memories.sessionId))
            .orderBy(desc(Memories.id))
            .limit(limit)
            .execute();

    return {
      result: memories.reverse().map(this.convertMemory),
    };
  }

  async create(
    memory: Pick<Memory, "content">,
    { context }: AgentInvokeOptions,
  ): Promise<{ result: Memory }> {
    const sessionId = (await this.options?.getSessionId?.(context)) ?? null;

    const db = await this.db;

    const id = v7();

    const [[result]] = await Promise.all([
      db
        .insert(Memories)
        .values({
          ...memory,
          id,
          content: memory.content,
          sessionId,
        })
        .returning()
        .execute(),
      this.options?.enableFTS &&
        db.run(
          sql`\
          insert into Memories_fts (id, content)
          values (${sql.param(id)}, ${sql.param(this.segment(stringify(memory.content)).join(" "))})`,
        ),
    ]);

    if (!result) throw new Error("Failed to create memory");

    return { result: this.convertMemory(result) };
  }

  protected segment(str: string): string[] {
    return (
      Array.from(new Intl.Segmenter(undefined, { granularity: "word" }).segment(str))
        .map((i) => i.segment)
        // Remove non-alphanumeric characters and trim whitespace
        .map((i) => i.replace(/[^\p{L}\p{N}\s]/gu, "").trim())
        .filter(Boolean)
    );
  }
}
