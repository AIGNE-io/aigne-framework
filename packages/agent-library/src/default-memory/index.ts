import {
  type AgentInvokeOptions,
  type AgentOptions,
  type Memory,
  MemoryAgent,
  type MemoryAgentOptions,
  MemoryRecorder,
  type MemoryRecorderInput,
  type MemoryRecorderOutput,
  MemoryRetriever,
  type MemoryRetrieverInput,
  type MemoryRetrieverOutput,
  type Message,
} from "@aigne/core";
import {
  DefaultMemoryStorage,
  type DefaultMemoryStorageOptions,
} from "./default-memory-storage/index.js";
import { MemoryStorage } from "./storage.js";

const DEFAULT_RETRIEVE_MEMORY_COUNT = 10;

export interface DefaultMemoryOptions extends Partial<MemoryAgentOptions> {
  storage?: MemoryStorage | DefaultMemoryStorageOptions;
  recorderOptions?: Omit<DefaultMemoryRecorderOptions, "storage">;
  retrieverOptions?: Omit<DefaultMemoryRetrieverOptions, "storage">;
}

export class DefaultMemory extends MemoryAgent {
  constructor(options: DefaultMemoryOptions = {}) {
    const storage =
      options.storage instanceof MemoryStorage
        ? options.storage
        : new DefaultMemoryStorage(options.storage);

    super({
      ...options,
      recorder:
        options.recorder ??
        new DefaultMemoryRecorder({
          ...options.recorderOptions,
          storage,
        }),
      retriever:
        options.retriever ??
        new DefaultMemoryRetriever({
          ...options.retrieverOptions,
          storage,
        }),
      autoUpdate: options.autoUpdate ?? true,
    });

    this.storage = storage;
  }

  storage: MemoryStorage;
}

export interface DefaultMemoryRetrieverOptions
  extends AgentOptions<MemoryRetrieverInput, MemoryRetrieverOutput> {
  storage: MemoryStorage;
  defaultRetrieveMemoryCount?: number;
}

class DefaultMemoryRetriever extends MemoryRetriever {
  constructor(options: DefaultMemoryRetrieverOptions) {
    super(options);
    this.storage = options.storage;
    this.defaultRetrieveMemoryCount = options.defaultRetrieveMemoryCount;
  }

  private storage: MemoryStorage;

  private defaultRetrieveMemoryCount?: number;

  async process(
    input: MemoryRetrieverInput,
    options: AgentInvokeOptions,
  ): Promise<MemoryRetrieverOutput> {
    const { result } = await this.storage.search(
      {
        ...input,
        limit: input.limit ?? this.defaultRetrieveMemoryCount ?? DEFAULT_RETRIEVE_MEMORY_COUNT,
      },
      options,
    );
    return { memories: result };
  }
}

export interface DefaultMemoryRecorderOptions
  extends AgentOptions<MemoryRecorderInput, MemoryRecorderOutput> {
  storage: MemoryStorage;
  rememberFromMessageKey?: string | string[];
}

class DefaultMemoryRecorder extends MemoryRecorder {
  constructor(options: DefaultMemoryRecorderOptions) {
    super(options);
    this.storage = options.storage;
    this.rememberFromMessageKey = options.rememberFromMessageKey;
  }

  private storage: MemoryStorage;

  private rememberFromMessageKey?: string | string[];

  async process(
    input: MemoryRecorderInput,
    options: AgentInvokeOptions,
  ): Promise<MemoryRecorderOutput> {
    const newMemories: Memory[] = [];

    for (const item of input.content) {
      const { result } = await this.storage.create(
        {
          content: {
            ...item,
            content: this.processMemoryInput(item.content),
          },
        },
        options,
      );
      newMemories.push(result);
    }

    return {
      memories: newMemories,
    };
  }

  private processMemoryInput(content: Message) {
    const keys = Array.isArray(this.rememberFromMessageKey)
      ? this.rememberFromMessageKey
      : this.rememberFromMessageKey
        ? [this.rememberFromMessageKey]
        : [];
    if (!keys.length) return content;

    return Object.entries(content)
      .filter(([key]) => keys.includes(key))
      .map(([_, value]) => (typeof value === "string" ? value : JSON.stringify(value)))
      .join("\n");
  }
}
