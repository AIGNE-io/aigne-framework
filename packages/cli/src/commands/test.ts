import { spawnSync } from "node:child_process";
import { isAbsolute, resolve } from "node:path";
import { Command } from "commander";

export function createTestCommand(): Command {
  return new Command("test")
    .description("Run tests in the specified agents directory")
    .option(
      "--url, --path <path_or_url>",
      "Path to the agents directory or URL to aigne project",
      ".",
    )
    .action(async (options: { path: string }) => {
      const { path } = options;
      const absolutePath = isAbsolute(path) ? path : resolve(process.cwd(), path);

      spawnSync("node", ["--test"], { cwd: absolutePath, stdio: "inherit" });
    })
    .showHelpAfterError(true)
    .showSuggestionAfterError(true);
}
