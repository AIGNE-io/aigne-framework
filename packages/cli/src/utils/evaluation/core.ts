import { Listr } from "@aigne/listr2";
import { ConsoleReporter } from "./reporter.js";
import type {
  Dataset,
  DatasetItem,
  Evaluation,
  EvaluationSummary,
  Evaluator,
  Report,
  Reporter,
  Runner,
  RunOptions,
  RunResult,
} from "./type.ts";

function aggregateSummary(
  results: (RunResult & { evaluations: Evaluation[]; error?: any })[],
): EvaluationSummary {
  const total = results.length;
  const scores = results.flatMap((r) => r.evaluations.map((e) => e.score));
  const successRate = Number(
    (scores.length > 0 ? scores.reduce((a, b) => a + b, 0) / scores.length : 0).toFixed(2),
  );

  const latencies = results.map((r) => r.latency || 0);
  const totalTokens = results.reduce(
    (a, r) => a + (r.usage?.inputTokens || 0) + (r.usage?.outputTokens || 0),
    0,
  );

  const errors = results.filter((r) => r.error).length;

  return {
    total,
    successRate,
    avgLatency: latencies.reduce((a, b) => a + b, 0) / (latencies.length || 1),
    maxLatency: Math.max(...latencies, 0),
    minLatency: Math.min(...latencies, 0),
    totalTokens,
    errorCount: errors,
    scoreDistribution: {
      min: Math.min(...scores, 0),
      max: Math.max(...scores, 0),
      mean: successRate,
      median: scores.length ? scores.sort((a, b) => a - b)[Math.floor(scores.length / 2)] : 0,
      variance:
        scores.length > 1
          ? scores.reduce((a, b) => a + (b - successRate) ** 2, 0) / scores.length
          : 0,
    },
  };
}

type EvaluationPipelineContext = {
  items: DatasetItem[];
  results: (RunResult & { evaluations: Evaluation[] })[];
};

export async function runEvaluationPipeline(params: {
  dataset: Dataset;
  runner: Runner;
  evaluators: Evaluator[];
  reporters?: Reporter[];
  options?: RunOptions;
}) {
  const { dataset, runner, evaluators, reporters = [new ConsoleReporter()], options } = params;

  const results: EvaluationPipelineContext["results"] = [];

  const task1 = new Listr<{ items: DatasetItem[] }>([
    {
      title: "Load dataset",
      task: async (ctx, _task) => {
        ctx.items = await dataset.loadWithOptions();
      },
    },
  ]);

  const { items } = await task1.run();

  const task2 = new Listr<EvaluationPipelineContext>(
    items.map((item) => {
      const input = JSON.stringify(item.input);
      return {
        title: `Run evaluations with input: ${input.length > 100 ? `${input.slice(0, 100)}...` : input}...`,
        task: async (ctx, task) => {
          task.output = `Start running agent with input: ${JSON.stringify(item.input, null, 2)}`;

          const runnerResults = runner.run([item], options);

          for await (const result of runnerResults) {
            task.output = `Start running evaluation with: ${JSON.stringify(
              {
                input: result.input,
                output: result.output,
                expected: result.expected,
              },
              null,
              2,
            )}`;

            const evaluations: Evaluation[] = [];
            for (const evaluator of evaluators) {
              const evals = await evaluator.evaluate(result);
              evaluations.push(...evals);
            }

            results.push({ ...result, evaluations });

            task.output = `Finish running evaluation`;
          }

          ctx.results = results;
        },
      };
    }),
    {
      concurrent: options?.concurrency ? Math.min(items.length, options?.concurrency) : false,
      exitOnError: true,
      rendererOptions: {
        collapseSubtasks: false,
      },
    },
  );

  await task2.run();

  const summary: EvaluationSummary = aggregateSummary(results);
  const report: Report = { dataset: dataset.name, results, summary };

  for (const reporter of reporters) {
    await reporter.report(report);
  }
}
