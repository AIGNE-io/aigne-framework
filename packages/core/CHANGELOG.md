## 1.1.0-beta.17 (2025-3-18)

- chore: add support for esm module


## [1.2.0](https://github.com/AIGNE-io/aigne-framework/compare/core-next-v1.1.0...core-next-v1.2.0) (2025-03-18)


### Features

* add Agent FunctionAgent AIAgent MCPAgent and ExecutionEngine ([4d2a5a1](https://github.com/AIGNE-io/aigne-framework/commit/4d2a5a1b3366b8f935f50a0937c2da6e49073348))
* add OrchestratorAgent in agent library ([25a5e9e](https://github.com/AIGNE-io/aigne-framework/commit/25a5e9e6c60d747c8bf484ac884b31dc02c14757))
* add sequential and parallel helper function ([a295697](https://github.com/AIGNE-io/aigne-framework/commit/a295697b5694754e02954fc5c7f382a3b219a3ab))
* add support for MCP resources ([1ded2fb](https://github.com/AIGNE-io/aigne-framework/commit/1ded2fbf222fa8984e75df0852ff384524f73b04))
* **prompt-builder:** support chat history in PromptBuilder ([6ca05f2](https://github.com/AIGNE-io/aigne-framework/commit/6ca05f28eddb683a4f1e228865f8bbf8a8e190f1))
* support run puppeteer example chat loop in terminal ([85ce7f8](https://github.com/AIGNE-io/aigne-framework/commit/85ce7f8de8b443c86e50815dd7bcab99f869c4ce))
* use PromptBuilder instead of string instructions ([e4cb2cb](https://github.com/AIGNE-io/aigne-framework/commit/e4cb2cb6baf4f9bcef390567a4a174e9246c29a3))


### Bug Fixes

* **AIAgent:** should pass both arguments (model generated) and input (user input) to tool ([c49d64e](https://github.com/AIGNE-io/aigne-framework/commit/c49d64ee35f7efd83b0f82f43205bb1c40f999e8))
* **MCP:** catch list resource error treat as empty list ([1885fab](https://github.com/AIGNE-io/aigne-framework/commit/1885fab3585e0dd1467b127e5b47cd0b98282bab))
* use text resource from MCP correctly ([8b9eba8](https://github.com/AIGNE-io/aigne-framework/commit/8b9eba83352ec096a2a5d4f410d4c4bde7420bce))

## 1.1.0-beta.16 (2025-3-18)

- chore: add puppeteer in linux need docker_container

## 1.1.0-beta.15 (2025-3-18)

- chore: make coverage report as text to terminal
- chore: update contributing docs

## 1.1.0-beta.14 (2025-3-18)

- chore(example): add code-execution example

## 1.1.0-beta.13 (2025-3-18)

- feat: add OrchestratorAgent in agent library

## 1.1.0-beta.12 (2025-3-14)

- chore(example): add concurrency reflection handoff workflow examples

## 1.1.0-beta.11 (2025-3-14)

- feat(core): add sequential and parallel helper function
- chore(examples): add workflow-sequential example

## 1.1.0-beta.10 (2025-3-13)

- chore: ensure required environment variables have values

## 1.1.0-beta.9 (2025-3-13)

- fix(MCP): catch list resource error treat as empty list

## 1.1.0-beta.8 (2025-3-13)

- fix(AIAgent): should pass both arguments (model generated) and input (user input) to tool
- chore(examples): add workflow-router example
- chore(examples): rename examples puppeteer-mcp-server and sqlite-mcp-server to mcp-server-puppeteer and mcp-server-sqlite

## 1.1.0-beta.7 (2025-3-13)

- chore: rename @aigne/core to @aigne/core-next

## 1.1.0-beta.6 (2025-3-13)

- chore(examples): default enable mcp debug message for examples

## 1.1.0-beta.5 (2025-3-13)

- feat: support chat history in PromptBuilder
- feat: add `prompts` for MCPAgent to consume prompts from MCP server
- chore: add sqlite-mcp-server example
- test: add more unit test cases

## 1.1.0-beta.4 (2025-3-12)

- feat: support run puppeteer example chat loop in terminal

## 1.1.0-beta.3 (2025-3-11)

- chore: set module type for core package

## 1.1.0-beta.2 (2025-3-11)

- feat: use PromptBuilder instead of string instructions
- refactor: use tools instead of skills
- chore(examples): add puppeteer-mcp-server example

## 1.1.0-beta.1 (2025-3-11)

- feat: add Agent FunctionAgent AIAgent MCPAgent and ExecutionEngine
