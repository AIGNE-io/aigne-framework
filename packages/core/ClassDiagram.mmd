classDiagram
    class PromptBuilderBuildOptions {
        +Context context
        +Agent agent
        +object input
        +ChatModel model
    }

    class Prompt {
        +List~object~ messages
        +List~Agent~ tools
        +object toolChoice
        +object responseFormat
    }

    PromptBuilder ..> PromptBuilderBuildOptions: dependency
    PromptBuilder ..> Prompt: dependency
    class PromptBuilder {
        +build(PromptBuilderBuildOptions options) Prompt
    }

    class Model {
        +call(Prompt input) object*
    }

    ChatModel --|> Model: inheritance
    class ChatModel {
    }

    ImageModel --|> Model: inheritance
    class ImageModel {
    }

    Agent --|> EventEmitter: inheritance
    class Agent {
        +string name
        +string description
        +object inputSchema
        +object outputSchema
        +List~string~ inputTopic
        +List~string~ nextTopic
        +List~Agent~ tools
        +Map~string, Agent~ skills

        +call(string input, Context context) object
        +call(object input, Context context) object

        +process(object input, Context context) object*
        -verifyInput() void
        -verifyOutput() void
    }

    AIAgent --|> Agent: inheritance
    AIAgent *.. PromptBuilder: composition
    class AIAgent {
        +ChatModel model
        +string instructions
        +string outputKey
        +object toolChoice
        +PromptBuilder promptBuilder
    }

    ImageAgent --|> Agent: inheritance
    ImageAgent *.. PromptBuilder: composition
    class ImageAgent {
        +ImageModel model
        +string instructions
        +PromptBuilder promptBuilder
    }

    FunctionAgent --|> Agent: inheritance
    class FunctionAgent {
        +Function fn
    }

    RPCAgent --|> Agent: inheritance
    class RPCAgent {
        +string url
    }

    MCPAgent --|> Agent: inheritance
    MCPAgent *.. MCPClient: composition
    class MCPAgent {
        +MCPClient client
    }

    class MCPClient {
    }

    class Message {
        +object output
    }

    MessageQueue ..> Message: dependency
    class MessageQueue {
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class History {
        +string id
        +string agentId
        +object input
        +object output
    }

    Context *.. ChatModel: composition
    Context *.. ImageModel: composition
    Context --|> MessageQueue: inheritance
    Context ..> History: dependency
    class Context {
        +ChatModel model
        +ImageModel imageModel
        +List~Agent~ tools

        +getHistories(string agentId, int limit) List~History~
        +addHistory(History history) void
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class EventEmitter {
        +on(): void
        +emit(): void
    }

    UserAgent --|> Agent: inheritance
    class UserAgent {
    }

    class ExecutionEngineRunOptions {
        +boolean concurrency
    }

    ExecutionEngine --|> Context: inheritance
    ExecutionEngine --|> EventEmitter: inheritance
    ExecutionEngine ..> UserAgent: dependency
    ExecutionEngine ..> ExecutionEngineRunOptions: dependency
    class ExecutionEngine {
        +run(Agent agent) UserAgent
        +run(string input) object
        +run(object input) object
        +run(object input, Agent ...agents) object
        +run(object input, ExecutionEngineRunOptions options, Agent ...agents) object
    }
