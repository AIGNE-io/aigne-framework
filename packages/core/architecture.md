## Class Definitions

```mermaid
classDiagram
    class PromptBuilderBuildOptions {
        +Context context
        +Agent agent
        +object input
        +ChatModel model
    }

    class Prompt {
        +List~object~ messages
        +List~Agent~ tools
        +object toolChoice
        +object responseFormat
    }

    PromptBuilder ..> PromptBuilderBuildOptions: dependency
    PromptBuilder ..> Prompt: dependency
    class PromptBuilder {
        +build(PromptBuilderBuildOptions options) Prompt
    }

    class Model {
        +call(Prompt input) object*
    }

    ChatModel --|> Model: inheritance
    class ChatModel {
    }

    ImageModel --|> Model: inheritance
    class ImageModel {
    }

    Agent --|> EventEmitter: inheritance
    class Agent {
        +string name
        +string description
        +object inputSchema
        +object outputSchema
        +List~string~ inputTopic
        +List~string~ nextTopic
        +List~Agent~ tools

        +call(string input, Context context) object
        +call(object input, Context context) object

        +process(object input, Context context) object*
        -verifyInput() void
        -verifyOutput() void
    }

    AIAgent --|> Agent: inheritance
    AIAgent *.. PromptBuilder: composition
    class AIAgent {
        +ChatModel model
        +string instructions
        +string outputKey
        +object toolChoice
        +PromptBuilder promptBuilder
    }

    ImageAgent --|> Agent: inheritance
    ImageAgent *.. PromptBuilder: composition
    class ImageAgent {
        +ImageModel model
        +string instructions
        +PromptBuilder promptBuilder
    }

    FunctionAgent --|> Agent: inheritance
    class FunctionAgent {
        +Function fn
    }

    RPCAgent --|> Agent: inheritance
    class RPCAgent {
        +string url
    }

    MCPAgent --|> Agent: inheritance
    MCPAgent *.. MCPClient: composition
    class MCPAgent {
        +MCPClient client
    }

    class MCPClient {
    }

    class Message {
        +object output
    }

    MessageQueue ..> Message: dependency
    class MessageQueue {
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class History {
        +string id
        +string agentId
        +object input
        +object output
    }

    Context *.. ChatModel: composition
    Context *.. ImageModel: composition
    Context --|> MessageQueue: inheritance
    Context ..> History: dependency
    class Context {
        +ChatModel model
        +ImageModel imageModel
        +List~Agent~ tools

        +getHistories(string agentId, int limit) List~History~
        +addHistory(History history) void
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    class EventEmitter {
        +on(): void
        +emit(): void
    }

    UserAgent --|> Agent: inheritance
    class UserAgent {
    }


    AIGNE --|> Context: inheritance
    AIGNE --|> EventEmitter: inheritance
    AIGNE ..> UserAgent: dependency
    class AIGNE {
        +call(Agent agent) UserAgent
        +call(Agent agent, string input) object
        +call(Agent agent, object input) object
    }
```
