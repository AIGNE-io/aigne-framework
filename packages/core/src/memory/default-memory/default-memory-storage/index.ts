import { type InferSelectModel, desc, eq, isNull } from "drizzle-orm";
import { drizzle } from "drizzle-orm/libsql";
import type { AgentInvokeOptions } from "../../../agents/agent.js";
import type { Context } from "../../../aigne/context.js";
import type { PromiseOrValue } from "../../../utils/type-utils.js";
import type { Memory } from "../../memory.js";
import { MemoryStorage } from "../storage.js";
import { migrate } from "./migrate.js";
import { Memories } from "./models/memory.js";

const DEFAULT_MAX_MEMORY_COUNT = 10;

export interface DefaultMemoryStorageOptions {
  url?: string;
  getSessionId?: (context: Context) => PromiseOrValue<string>;
}

export class DefaultMemoryStorage extends MemoryStorage {
  memories: {
    [groupId: string]: Memory[];
  } = {};

  constructor(public options?: DefaultMemoryStorageOptions) {
    super();
  }

  private _db: ReturnType<typeof this.initSqlite> | undefined;

  private async initSqlite() {
    const db = drizzle(this.options?.url ?? ":memory:");

    await migrate(db);

    return db;
  }

  get db() {
    this._db ??= this.initSqlite();

    return this._db;
  }

  private convertMemory(m: InferSelectModel<typeof Memories>): Memory {
    return {
      id: m.id,
      sessionId: m.sessionId,
      content: m.content,
      createdAt: m.createdAt.toISOString(),
    };
  }

  async search(
    query: { limit?: number },
    { context }: AgentInvokeOptions,
  ): Promise<{ result: Memory[] }> {
    const { limit = DEFAULT_MAX_MEMORY_COUNT } = query;

    const sessionId = (await this.options?.getSessionId?.(context)) ?? null;

    const db = await this.db;

    const memories = await db
      .select()
      .from(Memories)
      .where(sessionId ? eq(Memories.sessionId, sessionId) : isNull(Memories.sessionId))
      .orderBy(desc(Memories.id))
      .limit(limit)
      .execute();

    return {
      result: memories.reverse().map(this.convertMemory),
    };
  }

  async create(
    memory: Pick<Memory, "content">,
    { context }: AgentInvokeOptions,
  ): Promise<{ result: Memory }> {
    const sessionId = (await this.options?.getSessionId?.(context)) ?? null;

    const db = await this.db;

    const [result] = await db
      .insert(Memories)
      .values({
        ...memory,
        sessionId,
      })
      .returning()
      .execute();

    if (!result) throw new Error("Failed to create memory");

    return { result: this.convertMemory(result) };
  }
}
