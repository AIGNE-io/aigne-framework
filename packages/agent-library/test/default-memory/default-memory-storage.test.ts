import { expect, test } from "bun:test";
import { DefaultMemoryStorage } from "@aigne/agent-library/default-memory/default-memory-storage";
import { AIGNE } from "@aigne/core";

test("DefaultMemoryStorage should search by fts", async () => {
  const context = new AIGNE().newContext();

  const storage = new DefaultMemoryStorage({
    enableFTS: true,
  });
  await storage.create({ content: "我喜欢红色" }, { context });
  await storage.create({ content: "我讨厌绿色" }, { context });

  const { result } = await storage.search({ search: "绿色", limit: 10 }, { context });
  expect(result).toMatchSnapshot([
    {
      id: expect.any(String),
      createdAt: expect.any(String),
    },
  ]);
});
