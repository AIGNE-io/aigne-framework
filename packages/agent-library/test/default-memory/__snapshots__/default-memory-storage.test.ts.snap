// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultMemoryStorage should search by fts 1`] = `
[
  {
    "content": "我讨厌绿色",
    "createdAt": Any<String>,
    "id": Any<String>,
    "sessionId": null,
  },
]
`;
